crates/mcp-client/src/service.rs

```diff
@@ -0,0 +1,147 @@
+use std::future::Future;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio::sync::{mpsc, Mutex};
+use tower::Service;
+
+use crate::transport::{Error as TransportError, MessageRouter, Transport};
+use mcp_core::protocol::JsonRpcMessage;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ServiceError {
+    #[error("Transport error: {0}")]
+    Transport(#[from] TransportError),
+
+    #[error("Serialization error: {0}")]
+    Serialization(#[from] serde_json::Error),
+
+    #[error("Request timed out")]
+    Timeout(#[from] tower::timeout::error::Elapsed),
+
+    #[error("Transport not initialized")]
+    NotInitialized,
+
+    #[error("Transport already initialized")]
+    AlreadyInitialized,
+
+    #[error("Other error: {0}")]
+    Other(String),
+
+    #[error("Unexpected server response")]
+    UnexpectedResponse,
+}
+
+struct TransportServiceInner<T: Transport> {
+    transport: Arc<T>,
+    router: Mutex<Option<MessageRouter>>,
+    initialized: AtomicBool,
+}
+
+impl<T: Transport> TransportServiceInner<T> {
+    async fn ensure_initialized(&self) -> Result<MessageRouter, ServiceError> {
```

Comment:
Would it be worth exploring the use of the Type State Pattern here? We could get compile time guarantees about the state of the Transport and have no need for the `ensure_initialized` further down in `TransportService::call`'s method. (Disclaimer: i only recently learned about this pattern https://cliffle.com/blog/rust-typestate/)

Reply:
It would probably be possible to also use the Type State Pattern on the McpClient itself for more compile time guarantees

Reply:
this makes a lot of sense. i will defer this to another PR for now